File: apps/flutter_wallet/lib/src/core/native/parsers/network.dart
import '../../../models.dart';
import '../../../wallet_api.dart';
import 'fields.dart';

NetworkSettings networkSettingsFromNativeJson(Object? value) {
  if (value is! Map<String, dynamic>) {
    throw const WalletApiException('Invalid native response');
  }
  return NetworkSettings(
    chain: nativeStringField(value, 'chain'),
    networkName: nativeStringField(value, 'network_name'),
    chainId: optionalNativeString(value, 'chain_id'),
    enabled: nativeBoolField(value, 'enabled'),
    defaultRpcUrl: nativeStringField(value, 'default_rpc_url'),
    userRpcUrl: optionalNativeString(value, 'user_rpc_url'),
    indexerEndpoint: optionalNativeString(value, 'indexer_endpoint'),
    explorerUrl: optionalNativeString(value, 'explorer_url'),
    nativeSymbol: nativeStringField(value, 'native_symbol'),
    nativeDecimals: nativeIntField(value, 'native_decimals'),
  );
}

NetworkPrivacySettings networkPrivacySettingsFromNativeJson(Object? value) {
  if (value is! Map<String, dynamic>) {
    throw const WalletApiException('Invalid native response');
  }
  return NetworkPrivacySettings(
    proxyEnabled: nativeBoolField(value, 'proxy_enabled'),
    proxyMode: nativeStringField(value, 'proxy_mode'),
    proxyUrl: optionalNativeString(value, 'proxy_url'),
  );
}

File: crates/wallet_core/src/models/network.rs
use super::ChainId;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProxyMode {
    Custom,
    Tor,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ChainSettings {
    pub chain: ChainId,
    pub network_name: String,
    pub chain_id: Option<String>,
    pub enabled: bool,
    pub default_rpc_url: String,
    pub user_rpc_url: Option<String>,
    pub indexer_endpoint: Option<String>,
    pub explorer_url: Option<String>,
    pub native_symbol: String,
    pub native_decimals: u8,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct NetworkPrivacySettings {
    pub proxy_enabled: bool,
    pub proxy_mode: ProxyMode,
    pub proxy_url: Option<String>,
}
